z3/src/tactic.rs

```diff
@@ -0,0 +1,159 @@
+        Tactic {
+            ctx,
+            z3_tactic,
+        }
+    }
+
+    pub fn new(ctx: &'ctx Context, name: &'ctx str) -> Tactic<'ctx> {
+        let tactic_name = CString::new(name).unwrap();
+        Tactic {
+            ctx,
+            z3_tactic: unsafe {
+                let guard = Z3_MUTEX.lock().unwrap();
+                let t = Z3_mk_tactic(ctx.z3_ctx, tactic_name.as_ptr());
+                Z3_tactic_inc_ref(ctx.z3_ctx, t);
+                t
+            },
+        }
+    }
+
+    /// Return a tactic that just return the given goal.
+    pub fn create_skip(ctx: &'ctx Context) -> Tactic<'ctx> {
+        Tactic::new_from_z3(ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_skip(ctx.z3_ctx);
+            Z3_tactic_inc_ref(ctx.z3_ctx, t);
+            t
+        })
+    }
+
+    /// Return a tactic that always fails.
+    pub fn create_fail(ctx: &'ctx Context) -> Tactic<'ctx> {
+        Tactic::new_from_z3(ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_fail(ctx.z3_ctx);
+            Z3_tactic_inc_ref(ctx.z3_ctx, t);
+            t
+        })
+    }
+
+    /// Return a tactic that keeps applying `t` until the goal is not modified anymore or the maximum
+    /// number of iterations `max` is reached.
+    pub fn repeat(ctx: &'ctx Context, t: Tactic, max: u32) -> Tactic<'ctx> {
+        Tactic {
+            ctx,
+            z3_tactic: unsafe {
+                let guard = Z3_MUTEX.lock().unwrap();
+                let t = Z3_tactic_repeat(ctx.z3_ctx, t.z3_tactic, max);
+                Z3_tactic_inc_ref(ctx.z3_ctx, t);
+                t
+            },
+        }
+    }
+
+    /// Return a tactic that applies the current tactic to a given goal and
+    /// the `then_tactic` to every subgoal produced by the original tactic.
+    pub fn and_then(&self, then_tactic: Tactic) -> Tactic {
+        let t = unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_and_then(self.ctx.z3_ctx, self.z3_tactic, then_tactic.z3_tactic);
+            Z3_tactic_inc_ref(self.ctx.z3_ctx, t);
+            t
+        };
+        Tactic {
+            ctx: self.ctx,
+            z3_tactic: t,
+        }
+    }
+
+    /// Return a tactic that current tactic to a given goal,
+    /// if it fails then returns the result of `else_tactic` applied to the given goal.
+    pub fn or_else(&self, else_tactic: Tactic) -> Tactic {
+        let t = unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_or_else(self.ctx.z3_ctx, self.z3_tactic, else_tactic.z3_tactic);
+            Z3_tactic_inc_ref(self.ctx.z3_ctx, t);
+            t
+        };
+        Tactic {
+            ctx: self.ctx,
+            z3_tactic: t,
+        }
+    }
+
+    pub fn apply(self, goal: &Goal, params: &Params) -> Goal<'ctx> {
+        let result = unsafe {
+            Z3_tactic_apply(
```

Comment:
I don't see `params` being used here. If params is optional, could you make the signature `params: Option<&Params>`, match on that, and apply the arms `None => Z3_tactic_apply, Some(params) => Z3_tactic_apply_ex`?